### PR TITLE
[Keymap] Update bcat's keymaps/userspace to share logic, add OLED functionality, and set up one of my macropads for WFH

### PR DESCRIPTION
**Pull request by bcat**
## Description

There's a few changes here that I've been testing over the past several months and that I'd like to go ahead and merge before anything bit rots too much. I can split this into smaller PRs if that's preferable, just let me know.

~Unfortunately, my Corne keymap is getting a bit big, and is now oversized on avr-gcc 5.4.0 that Debian Bullseye ships with. I managed to squeak just under the limit (just 28 bytes to spare!) with an avr-gcc 8.5.0 Linux toolchain [built by Lumito](https://avr.lumito.net/), but will probably need something like `NO_ACTION_TAPPING` (with the layer fixes from #11528) if core grows any bigger. Or maybe I should just give up some RGB modes.... :)~

After merging in latest QMK changes, it seems I've got a fair number of bytes left for my Crkbd (712 bytes free with avr-gcc 8.5.0). :)

First, some organizational improvements and cleanups:

* Added a script (`users/bcat/compile.sh`) to my userspace to build all the keyboard/keymap combos I use. This makes it easy to test changes and make sure I'm not breaking anything, as well as to rebuild all my firmware binaries after pulling latest QMK master into my fork.
* Moved userspace RGB configuration into its own source file.
* Factored my standard layer configurations for ortho/columnar-staggered and row-staggered keyboards into my userspace to reduce duplication across keymaps. (The `BCAT_ORTHO_LAYERS` macro controls which set of layers is used.)
* Factored my standard keycode aliases into my userspace to further cut down on boilerplate.
* Disabled a variety of unused features to enable my Crkbd to build with OLED + RGB Matrix features enabled. 
* Removed `NK_TOGG` keybindings since I use `FORCE_NKRO` now. (It "just works" on all my hardware, even in BIOS.)
* Only activate Bootmagic Lite for keyboards that don't have a reset button accessible: Polaris (no reset hole on case), DZ60 (early rev. 3 PCBs put reset button in a nonstandard place), KBD67 Hotswap (no reset hole on case). This lets me save a few bytes on boards like the the Crkbd that do have accessible reset buttons.

Second, the addition of an OLED library to my userspace:

* Added a basic library of helpers for sharing the bulk of my `oled_task_user` implementation across keymaps. I added a few weak functions to my userspace following the pattern established by QMK core.
* Added a framework to my userspace for implementing "keyboard pets", simple animations of a few frames each that can respond to keystrokes, etc.
* This framework was originally inspired by the [Luna keyboard pet](https://www.reddit.com/r/olkb/comments/lmtgxc/introducing_luna_the_qmk_keyboard_pet/) by HellSingCoder [already in the repo](https://github.com/qmk/qmk_firmware/blob/6dfe915e26d7147e6c2bed495d3b01cf5b21e6ec/keyboards/sofle/keymaps/helltm/keymap.c). However, I didn't share any code with that implementation in the framework itself, but instead generalized the concept to support different animations and to better fit the particular structure of my keymaps.
* Implemented custom OLED timeout behavior since the QMK core implementation doesn't allow the OLED to ever turn off if an animation is continuously running.
* Use the new `SPLIT_OLED_ENABLE` feature to sync OLED on/off status set by the custom timeout impl between halves of the keyboard. :)

Third, implemented a couple of "keyboard pets" in my userspace framework using open-source artwork:

* Implemented "Luna" keyboard pet in my userspace using GPL'd artwork by @HellSingCoder from his original Luna implementation. I preserved his original copyright notice and included attribution, so it should be clear that this isn't the original implementation, but rather an adaption. (~~However, if HellSingCoder would rather I rename the `bcat_oled_pet_luna.c` file to something else to make it extra clear this isn't his canonical version, I'd be happy to do that too.~~ And [he's good with the naming too](https://github.com/qmk/qmk_firmware/pull/14702#issuecomment-933787803). :D)
* Implemented "Isda" keyboard pet in my userspace using GPL'd artwork from [OpenGameArt user sparrow666](https://opengameart.org/users/sparrow666). Once again I included attribution/license info, and in this case the name (`bcat_oled_pet_isda.c`) is just a reference to a D&D campaign I play in. :)

Finally, actual functional changes to keymaps:

* Wired up OLED for Lily58 and Crkbd, using the "Luna" keyboard pet. (The "Isda" keyboard pet is used by my Unicorne keymap, but that keyboard [isn't merged into upstream QMK yet](https://github.com/qmk/qmk_firmware/pull/12993), so I excluded the associated keymap from this PR.)
* Updated my 9-Key macropad's keyboard for use when working from home.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [X] New feature
- [X] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [X] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

None

## Checklist

- [X] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [X] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [X] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by HellSingCoder** [comment]
Good job! I'm happy for you to keep luna as a name 👍🏼

**Comment by bcat** [comment]
> Good job! I'm happy for you to keep luna as a name 👍🏼

Sounds good, thanks much for the idea & original implementation!

**Comment by HellSingCoder** [comment]
I wasn't expecting all the attention that it got to be honest but, I'm glad that a lot of people liked it. If you have a video of it working I'll be happy to see it 😄

File: users/bcat/rules.mk
Comment (drashna) [inline users/bcat/rules.mk:41]: This is a bit more complicated (well, maybe not just "a bit" ...) but this should be much more flexible.  Eg, you should be able to just drop more files into the folder, without needing to change anything. 

This is definitely optional, and should be tested if applied.

```suggestion
	BCAT_OLED_PET ?= no
	ifneq ($(strip $(BCAT_OLED_PET)), no)
		ifneq (,$(wildcard $(USER_PATH)/bcat_oled_pet_$(strip $(BCAT_OLED_PET)).c))
	        SRC += $(USER_PATH)/bcat_oled_pet_$(strip $(BCAT_OLED_PET)).c
			OPT_DEFS += -DBCAT_OLED_PET
			OPT_DEFS += -DBCAT_OLED_PET_$(strip $(shell echo $(BCAT_OLED_PET) | tr '[:lower:]' '[:upper:]'))
		endif
	endif
endif
```

**Comment by bcat** [comment]
Thanks for the review! BTW, the review email I got showed a change to `BCAT_OLED_PET` logic in `rules.mk`, but I don't see the suggestion in GitHub for some reason. Still want me to give that a try? I'm not opposed if there's a way to make that logic more general purpose.

**Comment by bcat** [comment]
I went ahead and incorporated a simplified version of the logic from the email. It doesn't explicitly check for a missing file in the makefile, but instead just lets it linker error like so:

```
Linking: .build/crkbd_rev1_bcat_split_3x6_3.elf                                                     [ERRORS]
 |
 | avr-gcc: error: .build/obj_crkbd_rev1_bcat_split_3x6_3/bcat_oled_pet_cow.o: No such file or directory
 |
```

I feel like the error is clear enough that that's fine in practice, though.

**Comment by drashna** [comment]
> Thanks for the review! BTW, the review email I got showed a change to `BCAT_OLED_PET` logic in `rules.mk`, but I don't see the suggestion in GitHub for some reason. Still want me to give that a try? I'm not opposed if there's a way to make that logic more general purpose.

I removed it, because I ran into issues with it, and ... yeah.

**Comment by drashna** [comment]
And it quick check looks like this may be what you want to actually check if the file exists. 
```make
ifneq ("$(wildcard $(USER_PATH)/bcat_oled_pet_$(BCAT_OLED_PET).c)","")
```

**Comment by bcat** [comment]
> And it quick check looks like this may be what you want to actually check if the file exists.
> 
> ```makefile
> ifneq ("$(wildcard $(USER_PATH)/bcat_oled_pet_$(BCAT_OLED_PET).c)","")
> ```

Do you mind if I leave this out? I feel like it's a little cleaner to have a build error if the `BCAT_OLED_PET` value doesn't correspond to a valid source file than to silently ignore invalid values.

But I don't feel strongly about this, so if you do have a preference I'll go with whatever you suggest.

**Comment by drashna** [comment]
Well, you can have it generate an error if it's not found, etc.   But this should be fine.  Mostly, it was to allow for a simple way to add more.

**Review by drashna — APPROVED** [review]

**Comment by bcat** [comment]
Rebased against latest master and resolved the merge conflict. Everything still builds okay. :)

**Comment by drashna** [comment]
Sorry, but again?

File: users/bcat/bcat_oled.c
Comment (drashna) [inline users/bcat/bcat_oled.c:132]: This can be replaced with `get_u8_str(wpm, ' ')`, now.

Comment (bcat) [reply users/bcat/bcat_oled.c]: Thanks, done!

**Comment by bcat** [comment]
> Sorry, but again?

No worries, all good. Everything should be re-merged now.

By the way, I also have a new keymap for the Unicorne keyboard ready: https://github.com/bcat/qmk_firmware/commit/5e4d65041e7081296a826bde6286dc56fc8fc5fc. I had left it out of this PR since the Unicorne keyboard was in various stages of "not merged yet" and "doesn't compile" for a bit. But now the Unicorne keyboard is working fine at HEAD, and I wasn't sure if I should add my keymap for it to this PR, or if it would be better to make a separate (much smaller) PR after this one is merged.

No preference on my end, and I'm not in any rush on this. :) Just figured I'd ask which option is preferable.

**Comment by drashna** [comment]
i'd say wait for it, or open up another pr.

**Comment by bcat** [comment]
Sounds good, I'll just wait till this is merged then. (Again, no rush from my end, and I know there's a big backlog of PRs. So I don't mind waiting.)

**Review by drashna — APPROVED** [review]

**Comment by bcat** [comment]
Just one more rebase atop master, no extra changes. Nothing to see here... :)

**Review by tzarc — APPROVED** [review]

**Pull request merged**
